docs/dependency-injection/dependency-injection.md

```diff
@@ -0,0 +1,206 @@
+# 1. What is Dependency Injection (DI)?
+
+Dependency injection (DI) is a design pattern that provide loosely-coupled, maintainable, and testable code. It improves modularity, maintainability, and testability, and is easily customizable and extensible.
```

Comment:
Let's add (on top of this) a small explanation what it means for the user:

- It allows to customise logic that is deep inside Oryx. 
- without this, it would mean large portions of the logic must be overridden
- without this, it would be tempting to create a lot of boilerplate, which we do not want
- this allows to override logic, while still be able to upgrade to newer versions of Oryx going forward

Reply:
let's add also:
- why we do not use a standard DI package
- our DI is vanilla JS and can be used in other frameworks

Reply:
part of the intro in next subsection, you might want to move it up maybe?

Reply:
This is a fairly complex topic, I'd prefer to write it from the most important using topics:

1. high-level intro on what is DI
2. Describe with code example how to provide a custom service using DI
3. Explain how to use DI in components
4. Explain how to use DI in service
5. Define custom services
6. Multi-providers